src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -1196,48 +1199,55 @@ public static void Reverse(ref byte buf, nuint length)
                     tempLast = Avx2.Permute2x128(tempLast, tempLast, 0b00_01);
 
                     // Store the reversed vectors
-                    tempLast.StoreUnsafe(ref buf, firstOffset);
-                    tempFirst.StoreUnsafe(ref buf, lastOffset);
-                }
-                buf = ref Unsafe.Add(ref buf, numIters * numElements);
-                length -= numIters * numElements * 2;
+                    tempLast.StoreUnsafe(ref first);
+                    tempFirst.StoreUnsafe(ref last);
+
+                    first = ref Unsafe.Add(ref first, (nuint)Vector256<byte>.Count);
+                    last = ref Unsafe.Subtract(ref last, (nuint)Vector256<byte>.Count);
+                    lastOffset -= (nuint)Vector256<byte>.Count * 2;
+                } while (lastOffset >= (nuint)Vector256<byte>.Count * 2);
```

Comment:
There's no reason to fall through to a remainder loop, nor do you need 2 full vectors worth of values to use the SIMD implementation.  It's perfectly fine if the vectors overlap.

Taking a 17-byte input as an example for SSE3:

```
input:  |0|1|2|3|4|5|6|7|8|9|a|b|c|d|e|f|g|

first:    |f|e|d|c|b|a|9|8|7|6|5|4|3|2|1|0|
last:   |g|f|e|d|c|b|a|9|8|7|6|5|4|3|2|1|

output: |g|f|e|d|c|b|a|9|8|7|6|5|4|3|2|1|0|
```

Likewise, 33 bytes would be enough for the AVX2 loop, although it would be worth testing to see where the perf benefit kicks in.

Reply:
If I understand you correctly, this would only work on sizes between 1 and 2 vectors. If your input is larger than 2 vectors you need to reverse from opposite sides, and you couldn't do the remainder using your method because you'd swap the same values twice.
I'm not sure if it's worth the trouble needed for something that only works for the 32-64 byte range (avx example)

Reply:
If you loop `while first <= last` you will never double swap because you'll have at most the last iteration with any parts overlapping.  It works for inputs as small as one vector length, with no upper limit.

Reply:
Sorry, I drew that wrong when I looked at it 🤦‍♂️.  You can still end up with less than one vector worth of values that need to be swapped.

You could change your loop condition to `while Unsafe.IsAddressLessThan(ref left, ref right)`, however, which would allow the loop to work when the input is less than 2 vectors in length and would eliminate the bookkeeping for `lastOffset`.

The alternative to a fallthrough for cases where the remainder after n loops is less than one vector would be to back up the center vector before you start and swap and write it once at the end.  Might still be cheaper than falling through to a separate loop, but it complicates the code a bit.

Reply:
I added overlapping for the range between 1-2 vectors. I'm not sure how to explain this in better words, but if you try to overlap anything that's already been reversed (like using a fall through), you'd end up with moving the overlapped part back to its original state. You have to make sure to only overlap elements that hasn't been reversed yet, which means just having `Unsafe.IsAddressLessThan(ref left, ref right)` is not sufficient.
I tried to make it work without having to duplicate any code. Thanks for the suggestion and sorry for the very late response 👍

Reply:
Right, you can't double-reverse any part, but the logic you've used with `needTwoParts` is a bit confusing.  As long as your `last` is never less than your `first`, you won't be double reversing.

To illustrate, let's assume you're working with `Vector128<int>`, and you have a `Span<int>` of length 13 (3.25 vectors).  On your first iteration, the refs start here (you start by subtracting `Vector<T>.Count` from `last`):

```
first
|
|0|1|2|3|4|5|6|7|8|9|A|B|C|
                  |
                  last
```

After reversing each vector, swapping their locations, and incrementing/decrementing the refs, you have this:

```
        first
        |
|C|B|A|9|4|5|6|7|8|3|2|1|0|
          |
          last
```

Although the two overlap, it's still safe to continue because nothing pointed to has already been processed.  This is true as long as `first <= last`.  After one more iteration you have:

```
                first
                |
|C|B|A|9|8|7|6|5|4|3|2|1|0|
  |
  last
```

And the loop ends because `first > last`.  You would now add `Vector<int>.Count` back to `last` (because you started by subtracting it), and check to see if you have remainder > 1.  In this case, you don't, so you're done.

For a case that does include remainder, consider a `Span<int>` of length 10:

```
first
|
|0|1|2|3|4|5|6|7|8|9|
            |
            last
```

After the first iteration, you end up with:

```
        first
        |
|9|8|7|6|4|5|3|2|1|0|
    |
    last
```

Loop ends because `first > last`, but after re-adding `Vector<int>.Count` to `last`, you find that you have 2 remaining elements:

```
        first
        |
|9|8|7|6|4|5|3|2|1|0|
            |
            last
```

Which does require remainder processing.  Note that if your remainder is only 1, it's already sorted, so you only have to handle remainder between `2` and `Vector<T>.Count - 1`.

Reply:
Ahh I see what you mean now. Thanks!